### PR TITLE
feat(module:date-picker): add nzShowWeekNumber property

### LINKED ISSUE #3764: Option to show week numbers on all datepickers
## What problem does this feature solve?
The week picker shows week numbers automatically. I would love to be able to show week numbers on all datepickers as well. Some users (especially in Sweden) are used to use calendars by week numbers. Don't show the week numbers by default, but a setting to toggle this would be great.

## What does the proposed API look like?
Add `[nzShowWeekNumber]` (`bool`) option to all date pickers.

### PR DESCRIPTION
**Pull request by UncleDave**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Week numbers are only shown in the week picker component.

Issue Number: Fixes #3764 

## What is the new behavior?

`[nzShowWeekNumber]` can control whether week numbers are shown in the date picker. Week numbers are still always shown in the week picker.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by UncleDave** [comment]
@wenqi73 I know you guys have a huge stack of open PRs, but do you have any sort of estimate for when this PR might be reviewed?

**Comment by wenqi73** [comment]
@UncleDave CI failed, you should make it run successfully first.

**Comment by UncleDave** [comment]
@wenqi73 I'm not familiar with your CI pipeline, but I don't see how my changes could have caused this failure, nor can I reproduce it locally, do you have any clues about it?

```
✖ Compiling with Angular sources in Ivy partial compilation mode.
components/core/testing/component-bed.ts:38:5 - error TS2322: Type 'TestBed' is not assignable to type 'TestBedStatic'.
  Type 'TestBed' provides no match for the signature 'new (...args: any[]): TestBed'.

38     bed,
       ~~~

  components/core/testing/component-bed.ts:13:3
    13   bed: TestBedStatic;
         ~~~
    The expected type comes from property 'bed' which is declared here on type 'ComponentBed<T>'

[07:33:54] 'library:build-zorro' errored after 44 s
[07:33:54] Error: Process failed with code 1
    at ChildProcess.<anonymous> (/home/vsts/work/1/s/scripts/gulp/util/task-helpers.ts:34:25)
    at ChildProcess.emit (node:events:513:28)
    at ChildProcess.emit (node:domain:552:15)
    at maybeClose (node:internal/child_process:1093:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)
    at Process.callbackTrampoline (node:internal/async_hooks:130:17)
```

**Comment by wenqi73** [comment]
@UncleDave Run `npm run test` and `npm run build` can reproduce this.

**Comment by UncleDave** [comment]
@wenqi73 I ran those before my commit, they both succeed for me.

**Comment by wenqi73** [comment]
@UncleDave Maybe you should merge the latest code of master and re-install dependencies?

**Comment by UncleDave** [comment]
@wenqi73 Is there no `package-lock.json` file? It looks like my build is now failing due to dependency type mismatches. You can also see that every other PR is now failing with the same issue that I originally had on this branch. Does this project have a solution for locking dependencies? It's very hard to make a change and have confidence it will build on the CI server.

File: package.json [hunk 1]
```diff
@@ -33,9 +33,9 @@
   },
   "devDependencies": {
     "@angular-builders/custom-webpack": "^14.0.0",
-    "@angular-devkit/build-angular": "^14.1.0",
-    "@angular-devkit/core": "^14.1.0",
-    "@angular-devkit/schematics": "^14.1.0",
```
Comment (wenqi73) [inline package.json:38]: Undo changes to this file.

Comment (UncleDave) [reply package.json:38]: Then the build won't pass due to `@schematics/angular@4.2.0` updating their `typescript` dependency to a version this library doesn't support. I've only changed this because there's no lock file, but I can't undo it without a solution. You can see the build failure caused by the mismatch [here](https://dev.azure.com/ng-zorro/NG-ZORRO/_build/results?buildId=4936&view=logs&j=8680be89-378c-5901-27b8-79bf589da266&t=7cc10bac-49f7-5748-d4ee-64732cd3a79d).

File: components/date-picker/style/panel.less [hunk 2]
```diff
@@ -538,11 +538,7 @@
   }
 
   // ====================== Date Panel ======================
-  &-date-panel {
-    .@{picker-prefix-cls}-body {
-      padding: @padding-xs @padding-sm;
-    }
```
Comment (wenqi73) [inline components/date-picker/style/panel.less:544]: Could you also undo the changes here? In general, we will not change the styles here, because it's from ant-design.

File: components/date-picker/lib/abstract-table.ts [hunk 3]
```diff
@@ -30,6 +30,7 @@ export abstract class AbstractTable implements OnInit, OnChanges {
   @Input() disabledDate?: (d: Date) => boolean;
   @Input() cellRender?: string | TemplateRef<Date> | FunctionProp<TemplateRef<Date> | string>;
   @Input() fullCellRender?: string | TemplateRef<Date> | FunctionProp<TemplateRef<Date> | string>;
+  @Input() selectWeek: boolean = false;
```
Comment (wenqi73) [inline components/date-picker/lib/abstract-table.ts:33]: `canSelectWeek` will be a better name here.

File: components/date-picker/style/panel.less (hunk 2 above)
Comment (UncleDave) [reply components/date-picker/style/panel.less:544]: This is to support week numbers. This style is still applied, just not when week numbers are shown. Without this change the week numbers will cause the dates to be cut off.

File: components/date-picker/doc/index.en-US.md [hunk 4]
```diff
@@ -78,6 +78,7 @@ The following APIs are shared by nz-date-picker, nz-range-picker.
 | `[nzShowTime]` | to provide an additional time selection | `object \| boolean` | [TimePicker Options](/components/time-picker/en#api) |
 | `[nzShowToday]` | whether to show 'Today' button | `boolean` | `true` |
 | `[nzShowNow]` | whether to show 'Now' button on panel when `nzShowTime` is set | `boolean` | `true` |
+| `[nzShowWeekNumber]` | whether to show the week number on each row (Only supported by date picker. Week picker always shows week numbers) | `boolean` | `false` |
```
Comment (wenqi73) [inline components/date-picker/doc/index.en-US.md:81]: Should also change on the Chinese file:
是否在每一行显示周数（仅日期选择器支持。周选择器始终显示周数）

File: components/date-picker/demo/week-number.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
+---
+order: 16
+title:
+  zh-CN:
+  en-US: Week Number
+---
+
+## zh-CN
+
+
```
Comment (wenqi73) [inline components/date-picker/demo/week-number.md:10]: 使用 `nzShowWeekNumber` 显示周数

**Review by wenqi73 — CHANGES_REQUESTED** [review]
I left some comments.

File: components/date-picker/style/panel.less (hunk 2 above)
Comment (wenqi73) [reply components/date-picker/style/panel.less:544]: If this is necessary, add the styles on patch.less while keep the origin file.

File: package.json (hunk 1 above)
Comment (wenqi73) [reply package.json:38]: @simplejason I see the recent PRs run failed. This change can resolve this, what do you think of this?

File: components/date-picker/demo/week-number.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
+---
+order: 16
+title:
+  zh-CN:
```
Comment (UncleDave) [inline components/date-picker/demo/week-number.md:4]: @wenqi73 I'm missing a translation for the title, could you help me out?

Comment (wenqi73) [reply components/date-picker/demo/week-number.md:4]: 显示周数

File: package.json (hunk 1 above)
Comment (wenqi73) [reply package.json:38]: @UncleDave Could you open another PR for fixing CI? And we merge that first

Comment (UncleDave) [reply package.json:38]: I also did what is in #7637, which I will remove from this PR in favour of that one.

Comment (UncleDave) [reply package.json]: Dependency changes moved to #7640 
This PR should build after #7637 & #7640 are merged.

**Review by wenqi73 — APPROVED** [review]
LGTM

**Comment by UncleDave** [comment]
@wenqi73 Is there anything else you need from me here? The code cov changes are weird since it says the coverage of `ellipsis.directive.ts` has changed, but I don't really see how. I can try to write a test for that if needed?

**Review by simplejason — APPROVED** [review]
LGTM
